c/tskit/genotypes.c

```diff
@@ -565,6 +572,69 @@ tsk_variant_decode(
     return ret;
 }
 
+int
+tsk_variant_restricted_copy(tsk_variant_t *self, tsk_variant_t *other)
+{
+    int ret = 0;
+    tsk_size_t total_len, offset, j;
+
+    /* Copy everything */
+    tsk_memcpy(other, self, sizeof(tsk_variant_t));
+    /* Tree sequence left as NULL and zero'd tree is a way of indicating this variant is
+     * fixed and cannot be further decoded. */
+    other->tree_sequence = NULL;
+    tsk_memset(&other->tree, sizeof(tsk_tree_t), 0);
+    other->traversal_stack = NULL;
+    other->samples = NULL;
+    other->sample_index_map = NULL;
+    other->alt_samples = NULL;
+    other->alt_sample_index_map = NULL;
+    other->user_alleles_mem = NULL;
+
+    other->genotypes = tsk_malloc(other->num_samples * sizeof(int32_t));
+    if (other->genotypes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memcpy(other->genotypes, self->genotypes, other->num_samples * sizeof(int32_t));
+
+    other->allele_lengths
+        = tsk_malloc(other->num_alleles * sizeof(*other->allele_lengths));
+    if (other->allele_lengths == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memcpy(other->allele_lengths, self->allele_lengths,
+        other->num_alleles * sizeof(*other->allele_lengths));
+
+    total_len = 0;
+    for (j = 0; j < other->num_alleles; j++) {
+        total_len += other->allele_lengths[j];
+    }
+    other->user_alleles_mem = tsk_malloc(total_len * sizeof(char *));
+    if (other->user_alleles_mem == NULL) {
```

Comment:
Could batch the mallocs together above and check the pointers == NULL at the same time. Saves a bit of repetition and reduces noise in the coverage stats.